consensus/src/experimental/buffer.rs

```diff
@@ -0,0 +1,198 @@
+use diem_crypto::HashValue;
+use std::collections::HashMap;
+
+pub struct LinkedItem<T: Hashable> {
+    // use option so we don't need T to be cloneable
+    elem: Option<T>,
+    // order is for find_element_by_key to have a starting position (similar to find_element)
+    order: u64,
+    next: Option<HashValue>,
+}
+
+pub type Link = Option<HashValue>;
+
+pub struct Buffer<T: Hashable> {
+    map: HashMap<HashValue, LinkedItem<T>>,
+    count: u64,
+    head: Link,
+    tail: Link,
+}
+
+/// This function compares two links if they are not None
+/// It returns true if and only if both aare not None and their keys are the same
+pub fn link_eq(link_a: &Link, link_b: &Link) -> bool {
+    link_a.is_some() && link_b.is_some() && link_a.as_ref().unwrap() == link_b.as_ref().unwrap()
+}
+
+impl<T: Hashable> Buffer<T> {
+    pub fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+            count: 0,
+            head: None,
+            tail: None,
+        }
+    }
+
+    pub fn get_head(&self) -> &Link {
+        &self.head
+    }
+
+    pub fn get_tail(&self) -> &Link {
+        &self.tail
+    }
+
+    pub fn push_back(&mut self, elem: T) {
+        self.count = self.count.checked_add(1).unwrap();
+        let t_hash = elem.hash();
+        self.map.insert(
+            t_hash,
+            LinkedItem {
+                elem: Some(elem),
+                order: self.count,
+                next: None,
+            },
+        );
+        if let Some(tail) = self.tail {
+            self.map.get_mut(&tail).unwrap().next = Some(t_hash);
+        }
+        self.tail = Some(t_hash);
+        self.head.get_or_insert(t_hash);
+    }
+
+    pub fn pop_front(&mut self) -> Option<T> {
+        self.head.take().map(|head| {
+            let item = self.map.get_mut(&head).unwrap();
+            let elem = item.elem.take();
+            self.head = item.next;
+            if self.head.is_none() {
+                // empty
+                self.tail = None;
+            }
+            elem.unwrap()
+        })
+    }
+
+    // utils - assuming item is not None
+    pub fn get_next(&self, link: &Link) -> Link {
+        self.map.get(link.as_ref().unwrap()).unwrap().next
+    }
+
+    pub fn get_elem(&self, link: &Link) -> &T {
+        self.map
+            .get(link.as_ref().unwrap())
+            .unwrap()
+            .elem
+            .as_ref()
+            .unwrap()
+    }
+
+    pub fn set_elem(&mut self, link: &Link, new_val: T) {
+        self.map
+            .get_mut(link.as_ref().unwrap())
+            .unwrap()
+            .elem
+            .replace(new_val);
+    }
+
+    pub fn take_elem(&mut self, link: &Link) -> T {
+        self.map
+            .get_mut(link.as_ref().unwrap())
+            .unwrap()
+            .elem
+            .take()
+            .unwrap()
+    }
+
+    pub fn find_elem<F: Fn(&T) -> bool>(&self, link: Link, compare: F) -> Link {
+        let mut current = link;
+        while link.is_some() {
```

Comment:
this should be while current.is_some()?

Reply:
we need to add tests as well (fill the TODO part)

Reply:
thanks!